File: program/triangle.cpp
//------------------------------------------------------------------------------
// triangle.cpp - содержит процедуры для работы с треугольником
//------------------------------------------------------------------------------

#define _CRT_SECURE_NO_WARNINGS
#include "stdio.h"
#include "math.h"
#include "triangle.h"

//------------------------------------------------------------------------------
// Ввод параметров треугольника из потока
void In(triangle& t, FILE* file) {
    fscanf(file, "%d", &t.x_1);
    fscanf(file, "%d", &t.y_1);
    fscanf(file, "%d", &t.x_2);
    fscanf(file, "%d", &t.y_2);
    fscanf(file, "%d", &t.x_3);
    fscanf(file, "%d", &t.y_3);
}

// Случайный ввод параметров треугольника
void InRnd(triangle& t) {
    int a, b, c, x, y;
    do {
        t.x_1 = Random();
        t.y_1 = Random();
        t.x_2 = Random();
        t.y_2 = Random();
        t.x_3 = Random();
        t.y_3 = Random();

        x = std::abs(t.x_1 - t.x_2);
        y = std::abs(t.y_1 - t.y_2);
        a = sqrt(x * x + y * y);

        x = std::abs(t.x_2 - t.x_3);
        y = std::abs(t.y_2 - t.y_3);
        b = sqrt(x * x + y * y);

        x = std::abs(t.x_3 - t.x_1);
        y = std::abs(t.y_3 - t.y_1);
        c = sqrt(x * x + y * y);
    } while (a >= b + c || b >= a + c || c >= a + b);
}

//------------------------------------------------------------------------------
// Вывод параметров треугольника в поток
void Out(triangle& t, FILE* file) {
    fprintf(file, "It is Triangle. A-point: x = %d, y = %d. B-point: x = %d, y = %d. C-point: x = %d, y = %d. Perimeter = %f. Color: ",
        t.x_1, t.y_1, t.x_2, t.y_2, t.x_3, t.y_3, Perimeter(t));
}

//------------------------------------------------------------------------------
// Вычисление периметра треугольника
double Perimeter(triangle& t) {
    int x, y;
    x = std::abs(t.x_1 - t.x_2);
    y = std::abs(t.y_1 - t.y_2);
    double a = sqrt(x * x + y * y);

    x = std::abs(t.x_2 - t.x_3);
    y = std::abs(t.y_2 - t.y_3);
    double b = sqrt(x * x + y * y);

    x = std::abs(t.x_3 - t.x_1);
    y = std::abs(t.y_3 - t.y_1);
    double c = sqrt(x * x + y * y);
    return a + b + c;
}

File: hw3/program/triangle.py
import math
import color
import random
from shape import *
# ----------------------------------------------


class Triangle(Shape):
    def __init__(self):
        self.x_1 = int(0)
        self.y_1 = int(0)
        self.x_2 = 0
        self.y_2 = 0
        self.x_3 = 0
        self.y_3 = 0
        self.color = 0

    def get_figure(self, data, i):
        self.color = int(data[i])
        self.x_1 = int(data[i+1])
        self.y_1 = int(data[i+2])
        self.x_2 = int(data[i+3])
        self.y_2 = int(data[i+4])
        self.x_3 = int(data[i+5])
        self.y_3 = int(data[i+6])

    def get_random_figure(self):
        self.x_1 = random.randint(1, 21)
        self.y_1 = random.randint(1, 21)
        self.x_2 = random.randint(1, 21)
        self.y_2 = random.randint(1, 21)
        self.x_3 = random.randint(1, 21)
        self.y_3 = random.randint(1, 21)

        x = abs(self.x_1 - self.x_2)
        y = abs(self.y_1 - self.y_2)
        a = math.sqrt(x * x + y * y)

        x = abs(self.x_2 - self.x_3)
        y = abs(self.y_2 - self.y_3)
        b = math.sqrt(x * x + y * y)

        x = abs(self.x_3 - self.x_1)
        y = abs(self.y_3 - self.y_1)
        c = math.sqrt(x * x + y * y)

        while a >= b + c or b >= a + c or c >= a + b:
            self.x_1 = random.randint(1, 21)
            self.y_1 = random.randint(1, 21)
            self.x_2 = random.randint(1, 21)
            self.y_2 = random.randint(1, 21)
            self.x_3 = random.randint(1, 21)
            self.y_3 = random.randint(1, 21)

            x = abs(self.x_1 - self.x_2)
            y = abs(self.y_1 - self.y_2)
            a = math.sqrt(x * x + y * y)

            x = abs(self.x_2 - self.x_3)
            y = abs(self.y_2 - self.y_3)
            b = math.sqrt(x * x + y * y)

            x = abs(self.x_3 - self.x_1)
            y = abs(self.y_3 - self.y_1)
            c = math.sqrt(x * x + y * y)

    def perimeter(self):
        x = abs(self.x_1 - self.x_2)
        y = abs(self.y_1 - self.y_2)
        a = math.sqrt(x * x + y * y)

        x = abs(self.x_2 - self.x_3)
        y = abs(self.y_2 - self.y_3)
        b = math.sqrt(x * x + y * y)

        x = abs(self.x_3 - self.x_1)
        y = abs(self.y_3 - self.y_1)
        c = math.sqrt(x * x + y * y)
        return a + b + c

    def write(self, output):
        output.write("Triangle: x1 = {}  y1 = {}, x2 = {}  y2 = {}, "
                     "x3 = {}  y3 = {}. Perimeter = {}. Color is {}"
                     .format(self.x_1, self.y_1, self.x_2,
                             self.y_2, self.x_3, self.y_3,
                             self.perimeter(), color.get_color(self.color)))
